### PR TITLE
Howto scale independently

### PR DESCRIPTION
**Pull request by ignasi35**
References #25

### CODE REVIEW

File: docs-source/docs/modules/how-to/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/ShoppingCart.scala [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  def init(system: ActorSystem[_]): Unit = {
+    val behaviorFactory: EntityContext[Command] => Behavior[Command] = { entityContext =>
+      val i = math.abs(entityContext.entityId.hashCode % tags.size)
+      val selectedTag = tags(i)
+      ShoppingCart(entityContext.entityId, selectedTag)
+    }
+    val entity: Entity[Command, ShardingEnvelope[Command]] =
+      Entity(EntityKey)(behaviorFactory).withRole("eventsourcing")
+    ClusterSharding(system).init(entity)
+  }
```
Comment (ignasi35) [inline docs-source/docs/modules/how-to/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/ShoppingCart.scala:28]: I copy/pasted the minimum necessary so the snippet would compile.

File: docs-source/docs/modules/how-to/pages/scale-independently.adoc [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+Let's see these two changes in more detail.
+
+==== Configure the sharding
+
+TODO
+
+==== Set up the role in some nodes
+
+TODO
```
Comment (ignasi35) [inline docs-source/docs/modules/how-to/pages/scale-independently.adoc:99]: Each of the two sections (read-side and write-side) fololow the same identical structure:

```
== Tuning the XYZ-Side
To scale the projections independently you need make two changes:
. configure the sharding, so it only uses nodes with a certain role
. set up the role in some nodes and deploy
Let's see these two changes in more detail.
==== Configure the sharding
TODO
==== Set up the role in some nodes
TODO
```

File: docs-source/docs/modules/how-to/pages/scale-independently.adoc [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+For example:
+[#node-for-eventsourcing]
+.Node dedicated to event sourcing
+[caption=""]
+image::shopping-microservices:example-entity.png[Node dedicated to event sourcing]
+
+This type of node, would not run the projections. Alternatively, we could also have nodes dedicated to only run projections and no persistent entities.
+
+== Tuning the Read-Side
```
Comment (ignasi35) [inline docs-source/docs/modules/how-to/pages/scale-independently.adoc:35]:
```suggestion
== Tuning the Write-Side
```

Ooops

File: docs-source/docs/modules/shopping-microservices/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/ShoppingCart.scala [hunk 4]
```diff
@@ -133,13 +135,17 @@ object ShoppingCart {
...
+    val behaviorFactory: EntityContext[Command] => Behavior[Command] = { entityContext =>
       val i = math.abs(entityContext.entityId.hashCode % tags.size)
       val selectedTag = tags(i)
       ShoppingCart(entityContext.entityId, selectedTag)
-    })
+    }
+    val entity: Entity[Command, ShardingEnvelope[Command]] = Entity(EntityKey)(behaviorFactory)
+    ClusterSharding(system).init(entity)
   }
+  // end::howto-read-side-without-role[]
```
Comment (ignasi35) [inline docs-source/docs/modules/shopping-microservices/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/ShoppingCart.scala:148]: I did a small refactor here so the difference was more apparent.  I've also added new tags prefixed `howto-` so it's clear this is used from elsewhere.
The other option is to duplicate this bit into the code under the `how-to/example/` folder so both evolve independently.

Each option has its problems: for example, if the how-to didn't directly used the code under `shopping-microservice/example`, when that code evolves the How-to should no longer cross-reference the tutorial.

When we include Java code, this problem will only get worse. I am not sure what the best option is, TBH.

File: docs-source/docs/modules/how-to/examples/shopping-cart-service-scala/build.sbt [hunk 5]
```diff
@@ -0,0 +1,62 @@
+// FIXME once akka 2.6.9 is released
```
Comment (ignasi35) [inline docs-source/docs/modules/how-to/examples/shopping-cart-service-scala/build.sbt:1]: I've copy/pasted all the `build.sbt` (except the akka-grpc plugin) to simplify it's maintenance.

Comment (ignasi35) [reply docs-source/docs/modules/how-to/examples/shopping-cart-service-scala/build.sbt:1]: I should probably bring in the akka-grpc plugin aswell...

**Review by ignasi35 — COMMENTED** [review]
notes for reviewers

File: docs-source/docs/modules/how-to/examples/shopping-cart-service-scala/src/main/resources/application-eventsourcing.conf [hunk 6]
```diff
@@ -0,0 +1,8 @@
+# tag::roles-in-conf[]
+include "application"
+akka{
+  cluster{
```
Comment (patriknw) [inline docs-source/docs/modules/how-to/examples/shopping-cart-service-scala/src/main/resources/application-eventsourcing.conf:4]: formatting: `akka {` and `cluster {`

File: docs-source/docs/modules/how-to/pages/scale-independently.adoc [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+
+[source]
+----
+... -Dakka.cluster.roles.0=eventsourcing ...
+----
+
+
+== Tuning the Read-Side
+
+To scale the projections independently you need make two changes:
```
Comment (patriknw) [inline docs-source/docs/modules/how-to/pages/scale-independently.adoc:92]: need *to*

File: docs-source/docs/modules/how-to/pages/scale-independently.adoc [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+To scale the persistent entities independently you need make two changes:
+
+. configure the sharding, so it only uses nodes with a certain role
+. set up the role in some nodes and deploy
+
+Let's see these two changes in more detail.
+
+=== Configure the sharding
+
+When you initialize the sharding of a persistent entity, as described in the xref:shopping-microservices:entity.adoc#_initialization[tutorial], you may specify a role. Take for example the final code on the initialization of the persistent entity (from the tutorial):
```
Comment (patriknw) [inline docs-source/docs/modules/how-to/pages/scale-independently.adoc:46]: I think we often use capital Sharding or (Cluster Sharding, or Akka Cluster Sharding) to highlight that it's a product module

**Review by patriknw — COMMENTED** [review]
very good, just a few small things

File: docs-source/docs/modules/how-to/pages/scale-independently.adoc (hunk 8 above)
Comment (ignasi35) [reply docs-source/docs/modules/how-to/pages/scale-independently.adoc:46]: Ah, good point. I'll add a few cross-links to the ref docs at the bottom aswell.

**Comment by ignasi35** [comment]
Pressed the wrong buttons. Sorry. Let me know if there's any pending comment.

File: docs-source/docs/modules/shopping-microservices/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/ShoppingCart.scala [hunk 9]
```diff
@@ -133,13 +135,17 @@ object ShoppingCart {
...
   def init(system: ActorSystem[_]): Unit = {
-    ClusterSharding(system).init(Entity(EntityKey) { entityContext =>
+    val behaviorFactory: EntityContext[Command] => Behavior[Command] = { entityContext =>
       val i = math.abs(entityContext.entityId.hashCode % tags.size)
       val selectedTag = tags(i)
       ShoppingCart(entityContext.entityId, selectedTag)
-    })
+    }
+    val entity: Entity[Command, ShardingEnvelope[Command]] = Entity(EntityKey)(behaviorFactory)
+    ClusterSharding(system).init(entity)
```
Comment (patriknw) [inline docs-source/docs/modules/shopping-microservices/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/ShoppingCart.scala:146]: @ignasi35 I'm not sure I like this change. It makes it look more (complicated) than it is. At least I'd like to keep the last two lines as one, to not show the `ShardingEnvelope`. ShardingEnvelope isn't exposed to normal usage in other places so unnecessary concept to introduce here.

```
ClusterSharding(system).init(Entity(EntityKey)(behaviorFactory))
```

Comment (ignasi35) [reply docs-source/docs/modules/shopping-microservices/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/ShoppingCart.scala:146]: Sure, I'll amend on a new PR.

Comment (ignasi35) [reply docs-source/docs/modules/shopping-microservices/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/ShoppingCart.scala:146]: Raised #161

**Pull request merged**